src/java.base/share/classes/com/sun/crypto/provider/AESCrypt.java

```diff
@@ -602,13 +602,6 @@ private void implDecryptBlock(byte[] in, int inOffset,
      * @exception InvalidKeyException  If the key is invalid.
      */
     private void makeSessionKey(byte[] k) throws InvalidKeyException {
-        if (k == null) {
-            throw new InvalidKeyException("Empty key");
-        }
-        if (!isKeySizeValid(k.length)) {
-             throw new InvalidKeyException("Invalid AES key length: " +
-                                           k.length + " bytes");
-        }
         int ROUNDS          = getRounds(k.length);
```

Comment:
You could have left the check in here and removed the duplicate check from `init()`. Since the key is not referenced by `init` unless the key is different from the last key, it seems cleaner to leave it here, and I think the check would be invoked fewer times if the same key is reused.

Reply:
Actually, NM, init still has to call MessageDigest.isEqual so eliminating keys of invalid length before that is probably more efficient.

Reply:
Good point.  Modified to use less checking.

If the key is null, the following condition could bypass the checking, and result in NPE.

`        if (!MessageDigest.isEqual(key, lastKey)) {
`

Although it is unlikely to happen as the caller should has already been checked that the key cannot be null, but the code logic here is not that clear to read.  In the new patch, I have the null checking in the init() method, and the validity checking in the makeSessionKey() method.

Reply:
> Actually, NM, init still has to call MessageDigest.isEqual so eliminating keys of invalid length before that is probably more efficient.

The key should be valid for common cases.  For valid key, it is more efficient to have the checking in makeSessionKey() as there is less checking.  For invalid key, it is more efficient to have the checking before calling MessageDigest.isEqual().  Exception itself is costly, I would prefer to have better performance for common cases (valid key).

I updated the patch before I read the comment.  Please let me know your preference.  I'm fine to rollback if you prefer the old patch.

Reply:
> If the key is null, the following condition could bypass the checking, and result in NPE.
> 
> ` if (!MessageDigest.isEqual(key, lastKey)) {`
> 
> Although it is unlikely to happen as the caller should has already been checked that the key cannot be null, but the code logic here is not that clear to read. In the new patch, I have the null checking in the init() method, and the validity checking in the makeSessionKey() method.

I agree, the provider layer should have rejected a null `Key` or a null `Key.getEncoded` prior to this, but best to not change what this code previously did (at least not for this change).

Reply:
> > Actually, NM, init still has to call MessageDigest.isEqual so eliminating keys of invalid length before that is probably more efficient.
> 
> The key should be valid for common cases. For valid key, it is more efficient to have the checking in makeSessionKey() as there is less checking. For invalid key, it is more efficient to have the checking before calling MessageDigest.isEqual(). Exception itself is costly, I would prefer to have better performance for common cases (valid key).
> 
> I updated the patch before I read the comment. Please let me know your preference. I'm fine to rollback if you prefer the old patch.

Yes, I think your current fix should be fine too. No need to rollback.

Reply:
Speaking of MessageDigest.isEqual, we don't need constant time comparison here. We could use Arrays.equals for some extra performance.

Reply:
Actually, never mind that. We need constant time comparison to avoid leaking information about differences between old and new key. Sorry for the noise.

Reply:
@djelinski If both styles (w/o constant-time operations) get used in the code, it may take time to analysis the potential secret leaking issues for code readers until there is a clear comment.  As may add additional human and maintenance cost, which may be as expensive as the computer cost, especially when something goes wrong.  So normally, I prefer to constant-time operations for secret informations, no matter if the operations expose to attacking surfaces or not.  Just my $.02.